File: core/src/main/java/dev/vibris/core/OwnedSourceTree.java
package dev.vibris.core;

import dev.vibris.protocol.v1.ErrorCode;

import java.io.IOException;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;

import static java.nio.file.LinkOption.NOFOLLOW_LINKS;

final class OwnedSourceTree {
    private static final long MAX_BYTES = 512L * 1024 * 1024;
    private static final int MAX_FILES = 100_000;

    private final Path pendingRoot;
    private OwnedPathIdentity pendingRootIdentity;

    OwnedSourceTree(Path pendingRoot) {
        this.pendingRoot = pendingRoot.toAbsolutePath().normalize();
    }

    Inspection inspect(String uuid) throws SourceRegistry.Failure {
        requireSafePendingRoot();
        Path directory = pendingRoot.resolve(uuid).normalize();
        if (!pendingRoot.equals(directory.getParent())) {
            throw new SourceRegistry.Failure(ErrorCode.INVALID_SOURCE_UUID, "Source UUID escapes the pending root.");
        }
        FileStats stats = scan(directory);
        return new Inspection(directory, stats.files, stats.bytes);
    }

    synchronized Ownership reserve(Path directory, long fileCount, long totalBytes)
        throws SourceRegistry.Failure {
        requireSafePendingRoot();
        FileStats stats = scan(directory);
        if (stats.files != fileCount || stats.bytes != totalBytes) {
            throw new SourceRegistry.Failure(
                ErrorCode.SOURCE_DIRECTORY_MISSING, "Source changed before ownership transfer.");
        }
        try {
            return new Ownership(pendingRootIdentity, OwnedPathIdentity.captureDirectory(directory));
        } catch (IOException exception) {
            throw new SourceRegistry.Failure(
                ErrorCode.SOURCE_CONTAINS_REPARSE_POINT, "Source changed before ownership transfer.");
        }
    }

    boolean stillOwned(Path directory, Ownership ownership) {
        return ownership.rootIdentity.matchesDirectory(pendingRoot) &&
            ownership.directoryIdentity.matchesDirectory(directory);
    }

    private FileStats scan(Path directory) throws SourceRegistry.Failure {
        if (!Files.isDirectory(directory, NOFOLLOW_LINKS) || Files.isSymbolicLink(directory)) {
            throw new SourceRegistry.Failure(
                ErrorCode.SOURCE_DIRECTORY_MISSING, "Prepared source directory is missing.");
        }
        FileStats stats = new FileStats();
        try {
            Files.walkFileTree(directory, new SimpleFileVisitor<>() {
                @Override
                public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes attributes) throws IOException {
                    requireOrdinary(path, attributes);
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult visitFile(Path path, BasicFileAttributes attributes) throws IOException {
                    requireOrdinary(path, attributes);
                    if (!attributes.isRegularFile()) throw new IOException("non-ordinary source entry");
                    stats.add(attributes.size());
                    return FileVisitResult.CONTINUE;
                }
            });
        } catch (IOException exception) {
            throw new SourceRegistry.Failure(
                ErrorCode.SOURCE_CONTAINS_REPARSE_POINT, "Prepared source is not an ordinary tree.");
        }
        if (stats.files == 0) {
            throw new SourceRegistry.Failure(ErrorCode.SOURCE_DIRECTORY_MISSING, "Prepared source is empty.");
        }
        return stats;
    }

    private synchronized void requireSafePendingRoot() throws SourceRegistry.Failure {
        Path current = pendingRoot;
        try {
            OwnedPathIdentity identity = OwnedPathIdentity.captureDirectory(pendingRoot);
            if (pendingRootIdentity == null) pendingRootIdentity = identity;
            if (!pendingRootIdentity.matchesDirectory(pendingRoot)) {
                throw new SourceRegistry.Failure(
                    ErrorCode.SOURCE_CONTAINS_REPARSE_POINT, "Pending source root identity changed.");
            }
            while (current != null) {
                OwnedPathIdentity.captureDirectory(current);
                current = current.getParent();
            }
        } catch (IOException exception) {
            ErrorCode code = Files.exists(pendingRoot, NOFOLLOW_LINKS)
                ? ErrorCode.SOURCE_CONTAINS_REPARSE_POINT
                : ErrorCode.SOURCE_DIRECTORY_MISSING;
            throw new SourceRegistry.Failure(code, "Pending source root is missing or unsafe.");
        }
    }

    static boolean delete(Path root) {
        try {
            Files.walkFileTree(root, new SimpleFileVisitor<>() {
                @Override
                public FileVisitResult preVisitDirectory(Path directory, BasicFileAttributes attributes)
                    throws IOException {
                    requireOrdinary(directory, attributes);
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult visitFile(Path file, BasicFileAttributes attributes) throws IOException {
                    requireOrdinary(file, attributes);
                    if (!attributes.isRegularFile()) throw new IOException("non-ordinary owned source entry");
                    Files.delete(file);
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult postVisitDirectory(Path directory, IOException failure) throws IOException {
                    if (failure != null) throw failure;
                    Files.delete(directory);
                    return FileVisitResult.CONTINUE;
                }
            });
            return true;
        } catch (IOException exception) {
            return false;
        }
    }

    private static void requireOrdinary(Path path, BasicFileAttributes attributes) throws IOException {
        if (Files.isSymbolicLink(path) || attributes.isSymbolicLink() || attributes.isOther()) {
            throw new IOException("link-like source entry");
        }
    }

    record Inspection(Path directory, long fileCount, long totalBytes) {
    }

    record Ownership(OwnedPathIdentity rootIdentity, OwnedPathIdentity directoryIdentity) {
    }

    private static final class FileStats {
        private long bytes;
        private int files;

        private void add(long size) throws IOException {
            if (++files > MAX_FILES || size > MAX_BYTES - bytes) throw new IOException("source limit exceeded");
            bytes += size;
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/OwnedSourceTree.kt
package dev.vibris.core

import dev.vibris.protocol.v2.ErrorCode
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.LinkOption.NOFOLLOW_LINKS
import java.nio.file.Path
import java.nio.file.SimpleFileVisitor
import java.nio.file.attribute.BasicFileAttributes
import java.security.MessageDigest

internal class OwnedSourceTree(
    pendingRoot: Path,
    private val maxBytes: Long = DEFAULT_MAX_BYTES,
    private val maxFiles: Int = DEFAULT_MAX_FILES,
) {
    private val pendingRoot = pendingRoot.toAbsolutePath().normalize()
    private var pendingRootIdentity: OwnedPathIdentity? = null

    init {
        require(maxBytes > 0) { "maxBytes must be positive" }
        require(maxFiles > 0) { "maxFiles must be positive" }
    }

    @Throws(SourceRegistry.Failure::class)
    fun inspect(uuid: String): Inspection {
        requireSafePendingRoot()
        val directory = pendingRoot.resolve(uuid).normalize()
        if (pendingRoot != directory.parent) {
            throw SourceRegistry.Failure(ErrorCode.ERROR_CODE_INVALID_SOURCE, "Source UUID escapes the pending root.")
        }
        val stats = scan(directory)
        return Inspection(directory, stats.files.toLong(), stats.bytes)
    }

    @Synchronized
    @Throws(SourceRegistry.Failure::class)
    fun reserve(directory: Path, fileCount: Long, totalBytes: Long): Reservation {
        requireSafePendingRoot()
        val stats = scan(directory)
        if (stats.files.toLong() != fileCount || stats.bytes != totalBytes) {
            throw SourceRegistry.Failure(
                ErrorCode.ERROR_CODE_INVALID_SOURCE,
                "Source changed before ownership transfer.",
            )
        }
        return try {
            Reservation(
                Ownership(
                    checkNotNull(pendingRootIdentity),
                    OwnedPathIdentity.captureDirectory(directory),
                ),
                sha256(stats, directory),
            )
        } catch (_: IOException) {
            throw SourceRegistry.Failure(
                ErrorCode.ERROR_CODE_SOURCE_CONTAINS_REPARSE_POINT,
                "Source changed before ownership transfer.",
            )
        }
    }

    fun stillOwned(directory: Path, ownership: Ownership): Boolean {
        return ownership.rootIdentity.matchesDirectory(pendingRoot) &&
            ownership.directoryIdentity.matchesDirectory(directory)
    }

    @Throws(SourceRegistry.Failure::class)
    fun matchesSnapshot(directory: Path, snapshotSha256: String): Boolean {
        val stats = scan(directory)
        return sha256(stats, directory) == snapshotSha256
    }

    @Throws(SourceRegistry.Failure::class)
    private fun scan(directory: Path): FileStats {
        if (!Files.isDirectory(directory, NOFOLLOW_LINKS) || Files.isSymbolicLink(directory)) {
            throw SourceRegistry.Failure(
                ErrorCode.ERROR_CODE_INVALID_SOURCE,
                "Prepared source directory is missing.",
            )
        }
        val stats = FileStats(maxBytes, maxFiles)
        try {
            Files.walkFileTree(
                directory,
                object : SimpleFileVisitor<Path>() {
                    override fun preVisitDirectory(
                        path: Path,
                        attributes: BasicFileAttributes,
                    ): FileVisitResult {
                        requireOrdinary(path, attributes)
                        return FileVisitResult.CONTINUE
                    }

                    override fun visitFile(path: Path, attributes: BasicFileAttributes): FileVisitResult {
                        requireOrdinary(path, attributes)
                        if (!attributes.isRegularFile) {
                            throw IOException("non-ordinary source entry")
                        }
                        stats.add(path, attributes.size())
                        return FileVisitResult.CONTINUE
                    }
                },
            )
        } catch (_: IOException) {
            throw SourceRegistry.Failure(
                ErrorCode.ERROR_CODE_SOURCE_CONTAINS_REPARSE_POINT,
                "Prepared source is not an ordinary tree.",
            )
        }
        if (stats.files == 0) {
            throw SourceRegistry.Failure(ErrorCode.ERROR_CODE_INVALID_SOURCE, "Prepared source is empty.")
        }
        return stats
    }

    @Throws(SourceRegistry.Failure::class)
    private fun sha256(stats: FileStats, directory: Path): String = try {
        stats.sha256(directory)
    } catch (_: IOException) {
        throw SourceRegistry.Failure(
            ErrorCode.ERROR_CODE_SOURCE_CONTAINS_REPARSE_POINT,
            "Prepared source changed while its snapshot hash was computed.",
        )
    }

    @Synchronized
    @Throws(SourceRegistry.Failure::class)
    private fun requireSafePendingRoot() {
        var current: Path? = pendingRoot
        try {
            val identity = OwnedPathIdentity.captureDirectory(pendingRoot)
            if (pendingRootIdentity == null) {
                pendingRootIdentity = identity
            }
            if (!checkNotNull(pendingRootIdentity).matchesDirectory(pendingRoot)) {
                throw SourceRegistry.Failure(
                    ErrorCode.ERROR_CODE_SOURCE_CONTAINS_REPARSE_POINT,
                    "Pending source root identity changed.",
                )
            }
            while (current != null) {
                OwnedPathIdentity.captureDirectory(current)
                current = current.parent
            }
        } catch (_: IOException) {
            val code = if (Files.exists(pendingRoot, NOFOLLOW_LINKS)) {
                ErrorCode.ERROR_CODE_SOURCE_CONTAINS_REPARSE_POINT
            } else {
                ErrorCode.ERROR_CODE_INVALID_SOURCE
            }
            throw SourceRegistry.Failure(code, "Pending source root is missing or unsafe.")
        }
    }

    data class Inspection(
        val directory: Path,
        val fileCount: Long,
        val totalBytes: Long,
    ) {
        fun directory(): Path = directory

        fun fileCount(): Long = fileCount

        fun totalBytes(): Long = totalBytes
    }

    data class Reservation(
        val ownership: Ownership,
        val snapshotSha256: String,
    )

    data class Ownership(
        val rootIdentity: OwnedPathIdentity,
        val directoryIdentity: OwnedPathIdentity,
    ) {
        fun rootIdentity(): OwnedPathIdentity = rootIdentity

        fun directoryIdentity(): OwnedPathIdentity = directoryIdentity
    }

    private class FileStats(private val maxBytes: Long, private val maxFiles: Int) {
        var bytes = 0L
        var files = 0
        private val paths = ArrayList<Path>()

        @Throws(IOException::class)
        fun add(path: Path, size: Long) {
            files++
            if (files > maxFiles || size > maxBytes - bytes) {
                throw IOException("source limit exceeded")
            }
            bytes += size
            paths.add(path)
        }

        @Throws(IOException::class)
        fun sha256(root: Path): String {
            val digest = MessageDigest.getInstance("SHA-256")
            digest.update("vibris-source-tree-v1\u0000".toByteArray(Charsets.UTF_8))
            paths.sortedBy { root.relativize(it).toString().replace('\\', '/') }.forEach { path ->
                val relative = root.relativize(path).toString().replace('\\', '/').toByteArray(Charsets.UTF_8)
                digest.update('F'.code.toByte())
                digest.update(ByteBuffer.allocate(Int.SIZE_BYTES).putInt(relative.size).array())
                digest.update(relative)
                val before = Files.readAttributes(path, BasicFileAttributes::class.java, NOFOLLOW_LINKS)
                requireOrdinary(path, before)
                digest.update(ByteBuffer.allocate(Long.SIZE_BYTES).putLong(before.size()).array())
                Files.newInputStream(path).use { input ->
                    val buffer = ByteArray(1024 * 1024)
                    while (true) {
                        val count = input.read(buffer)
                        if (count < 0) break
                        digest.update(buffer, 0, count)
                    }
                }
                val after = Files.readAttributes(path, BasicFileAttributes::class.java, NOFOLLOW_LINKS)
                requireOrdinary(path, after)
                if (before.size() != after.size() || before.lastModifiedTime() != after.lastModifiedTime() ||
                    before.fileKey() != after.fileKey()) {
                    throw IOException("source changed while hashing")
                }
            }
            return digest.digest().joinToString("") { byte -> "%02x".format(byte) }
        }
    }

    companion object {
        private const val DEFAULT_MAX_BYTES = 512L * 1024 * 1024
        private const val DEFAULT_MAX_FILES = 100_000

        @JvmStatic
        fun delete(root: Path): Boolean {
            return try {
                Files.walkFileTree(
                    root,
                    object : SimpleFileVisitor<Path>() {
                        override fun preVisitDirectory(
                            directory: Path,
                            attributes: BasicFileAttributes,
                        ): FileVisitResult {
                            requireOrdinary(directory, attributes)
                            return FileVisitResult.CONTINUE
                        }

                        override fun visitFile(file: Path, attributes: BasicFileAttributes): FileVisitResult {
                            requireOrdinary(file, attributes)
                            if (!attributes.isRegularFile) {
                                throw IOException("non-ordinary owned source entry")
                            }
                            Files.delete(file)
                            return FileVisitResult.CONTINUE
                        }

                        override fun postVisitDirectory(directory: Path, failure: IOException?): FileVisitResult {
                            if (failure != null) {
                                throw failure
                            }
                            Files.delete(directory)
                            return FileVisitResult.CONTINUE
                        }
                    },
                )
                true
            } catch (_: IOException) {
                false
            }
        }

        @Throws(IOException::class)
        private fun requireOrdinary(path: Path, attributes: BasicFileAttributes) {
            if (Files.isSymbolicLink(path) || attributes.isSymbolicLink || attributes.isOther) {
                throw IOException("link-like source entry")
            }
        }
    }
}